### ISSUE TITLE
docker run stuck when daemon can't umount/remove container

### ISSUE THREAD

**Issue by kolyshkin** [moby/moby#34945]
**Description**

I was investigating some other issue and wrote a test script to reproduce it:
```shell
#!/bin/bash

set -e
set -u
set -o pipefail

TOTAL_CTS=100

IMG=busybox
CMD="sleep 0.2"
RUN="docker run --rm --net=none $IMG $CMD"

# load image, preheat
$RUN

time for i in $(seq 1 "$TOTAL_CTS"); do
	printf "Iter: %5d Jobs: %5d\r" "$i" "$(jobs -p | wc -l)"
	$RUN &
done

echo "Waiting..."
time wait
```

The script is supposed to run 100 short-lived containers via docker CLI and wait for all `docker` children to finish. The problem is (it's a separate issue), if docker daemon can't unmount/remove a container, the corresponding docker cmdline is stuck (without reporting any error).

Docker daemon error (NOT the subject of this bug):
```
Sep 22 19:23:20 kir-ce73-gd dockerd[31494]: time="2017-09-22T19:23:20.322977590Z" level=error msg="Error removing mounted layer 5b920fbba5c5f330a0cb5ec3a296bc31cedaddeddb5ee0189836f6a08dae7d74: remove /var/lib/docker/overlay2/228429157bed9afebfb6f2d9c0f15c0fc1e43cf41f621fbbf71b2f11db2f7119/merged: device or resource busy"
Sep 22 19:23:20 kir-ce73-gd dockerd[31494]: time="2017-09-22T19:23:20.323097734Z" level=error msg="error removing container" container=5b920fbba5c5f330a0cb5ec3a296bc31cedaddeddb5ee0189836f6a08dae7d74 error="driver \"overlay2\" failed to remove root filesystem for 5b920fbba5c5f330a0cb5ec3a296bc31cedaddeddb5ee0189836f6a08dae7d74: remove /var/lib/docker/overlay2/228429157bed9afebfb6f2d9c0f15c0fc1e43cf41f621fbbf71b2f11db2f7119/merged: device or resource busy"
```

When the above happens, process `docker run ...` is just stuck sitting there forever.

This was tested with both `overlay2` and `overlay` driver, and I guess this is the same with other graph drivers.

**Steps to reproduce the issue:**
1. On a clean CentOS 7.4 system, install docker-ce 17.06
2. `systemctl start docker`.
3. `sysctl fs.may_detach_mounts=0` (this is to reproduce the issue of not being able to umount container)
4. Run the above shell script.

**Describe the results you received:**
One or more `docker run` commands stuck forever

**Describe the results you expected:**
`docker run` exits with an error.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
[root@kir-ce73-gd ~]# docker version
Client:
 Version:      17.06.1-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   874a737
 Built:        Thu Aug 17 22:53:49 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.1-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   874a737
 Built:        Thu Aug 17 23:01:50 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
root@kir-ce73-gd ~]# docker info
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 1
Server Version: 17.06.1-ce
Storage Driver: overlay
 Backing Filesystem: extfs
 Supports d_type: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 6e23458c129b551d5c9871e5174f6b1b7f6d1170
runc version: 810190ceaa507aa2727d7ae6f4790c76ec150bd2
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-693.2.2.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.796GiB
Name: kir-ce73-gd
ID: DF43:5A5K:WZRF:7OS7:BMGM:YPKR:IQVI:UVML:APS7:5ZME:KM2Y:RYG7
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
A physical server running centos 7.3 upgraded to centos 7.4

**Comment by kolyshkin** [comment]
Can I self-assign it? I'm planning to work on it.

**Comment by kolyshkin** [comment]
now, this might be a bug in either Moby or in docker-cli, I don't know yet so filed it for Moby.

**Comment by kolyshkin** [comment]
Reproduced with docker-ce-17.09.0.ce-0.3.rc3.el7.centos.x86_64.

Can't reproduce with docker-ce-17.03.0. Looks like a recent regression.

**Comment by kolyshkin** [comment]
OK, docker-ce-17.05.0.ce-1.el7.centos works fine (no stuck issue), while docker-ce-17.06.0.ce-0.1.rc1 already has this bug.

Bug is also reproducible when using 17.05 client with 17.06 daemon, so it seems it's a daemon bug.

**Comment by thaJeztah** [comment]
@kolyshkin is this a duplicate of https://github.com/moby/moby/issues/34672 ? See the analysis on that one, and the PR's linked from there

**Comment by kolyshkin** [comment]
With docker-ce 17.05 (last version which doesn't have this bug) I see that `docker run` receives this from the daemon:
```
15516 20:33:55.483981 read(3, "153\r\n{\"status\":\"die\",\"id\":\"27a1b11f02ff2846d2b20eb88c2e0e462b6785db8802be6a59595c607658ff25\",\"from\":\"busybox\",\"Type\":\"container\",\"Action\":\"die\",\"Actor\":{\"ID\":\"27a1b11f02ff2846d2b20eb88c2e0e462b6785db8802be6a59595c607658ff25\",\"Attributes\":{\"exitCode\":\"0\",\"image\":\"busybox\",\"name\":\"happy_ride\"}},\"time\":1506112435,\"timeNano\":1506112435480676298}\n\r\n", 4096) = 346
15516 20:33:55.484321 read(3, 0xc420407000, 4096) = -1 EAGAIN (Resource temporarily unavailable)
15149 20:33:55.488340 read(5, "", 32777) = 0
15149 20:33:55.629880 read(3, "14c\r\n{\"status\":\"destroy\",\"id\":\"27a1b11f02ff2846d2b20eb88c2e0e462b6785db8802be6a59595c607658ff25\",\"from\":\"busybox\",\"Type\":\"container\",\"Action\":\"destroy\",\"Actor\":{\"ID\":\"27a1b11f02ff2846d2b20eb88c2e0e462b6785db8802be6a59595c607658ff25\",\"Attributes\":{\"image\":\"busybox\",\"name\":\"happy_ride\"}},\"time\":1506112435,\"timeNano\":1506112435625986813}\n\r\n", 4096) = 339
```
and exits.

Newer daemon never sends anything.

**Comment by kolyshkin** [comment]
Related to "updated Container Wait API functionality"? @jlhawn can you please take a look?

**Comment by jlhawn** [comment]
The old version used the `events` API to wait for the `destroy` event on the container when you use the `--rm` flag. 17.06 switched to using an updated `ContainerWait` API to wait for the condition that the container is removed.

The `wait` API call will unblock as soon as `container.SetRemoved()` is called here: https://github.com/moby/moby/blob/c982ee805d895a4fc2bf01f2d1d8afa06b408dd0/daemon/delete.go#L141-L143

Notice that it is immediately before also sending a `destroy` event.

But it looks like it would never get to execute that line because it errors earlier in that same function: https://github.com/moby/moby/blob/c982ee805d895a4fc2bf01f2d1d8afa06b408dd0/daemon/delete.go#L122

So, since the container is never actually removed, the `removed` wait-condition is never met, so that API call continues to block.

**Comment by kolyshkin** [comment]
@jlhawn thank you for detailed analysis! What would you suggest to fix the issue? A somewhat naive idea I have is to log the error and keep going instead of returning an error.

**Comment by kolyshkin** [comment]
Here's my naive approach; I have tested it works (i.e. `docker run --rm` now exits).

```diff
diff --git a/daemon/delete.go b/daemon/delete.go
index 99b515d..bfa0477 100644
--- a/daemon/delete.go
+++ b/daemon/delete.go
@@ -119,12 +119,12 @@ func (daemon *Daemon) cleanupContainer(container *container.Container, forceRemo
                metadata, err := daemon.stores[container.Platform].layerStore.ReleaseRWLayer(container.RWLayer)
                layer.LogReleaseMetadata(metadata)
                if err != nil && err != layer.ErrMountDoesNotExist && !os.IsNotExist(errors.Cause(err)) {
-                       return errors.Wrapf(err, "driver %q failed to remove root filesystem for %s", daemon.GraphDriverName(container.Platform), container.ID)
+                       logrus.Errorf("Driver %q can't remove root filesystem for %s: %v", daemon.GraphDriverName(container.Platform), container.ID, err)
                }
        }
 
        if err := system.EnsureRemoveAll(container.Root); err != nil {
-               return errors.Wrapf(err, "unable to remove filesystem for %s", container.ID)
+               logrus.Errorf("Unable to remove filesystem for %s: %v", container.ID, err)
        }
 
        linkNames := daemon.linkIndex.delete(container)
```

**Comment by jlhawn** [comment]
@kolyshkin While it's certainly okay to trigger the release of the `wait` API call, I am not familiar enough with this part of the daemon to know if the rest of the function would be okay if the errors were ignored. It looks like if it continued anyway, it would remove all of the deamon's internal references to the container as well. Today, it continues to track the container in a "dead" state where you can retry later to remove it (though I'm not sure about the success rate of retrying).

If we just want to unblock the `wait` API call, then moving the `container.SetRemoved()` to just after this line: https://github.com/moby/moby/blob/c982ee805d895a4fc2bf01f2d1d8afa06b408dd0/daemon/delete.go#L106
might be okay. It should be something like `defer container.SetRemoved()` though. You would probably also need to add extra logic to `container/state.go` to make sure that the `waitRemove` channel is not closed more than once (which would panic) probably with a `sync.Once` wrapper.

I'm a bit uncomfortable with that still though because the wait condition is `removed`. To make it a bit more well-rounded, it may be appropriate to add a new wait condition named `dead` and have `docker run` wait on that condition instead.

**Comment by jlhawn** [comment]
Oh and also have `docker run --rm ...` check to make sure that the container was actually removed (if not in detached mode). I would think you'd want to know whether or not the removal actually worked and report an error to the CLI user if not.

**Comment by thaJeztah** [comment]
@kolyshkin I think that code may actually have been added to prevent `docker container rm -f` from deleting a container from memory if the root filesystem failed to be removed. 

In older versions of docker, the error was removed, and resulted in orphaned container-filesystems being left behind on the host, with no way to remove them (other than removing `/var/lib/docker`); docker 17.06 keeps the container around so that another attempt to remove it can be done later on.

ping @cpuguy83 ^^

**Comment by kolyshkin** [comment]
From my perspective, we need to be able to return an error from ContainerWait in case container removal failed, so a client can do something about it -- report an error, also maybe retry removal, or retry waiting for container to be removed.

As daemon returns http code 200 immediately and the error reveals itself later, I don't see any other way of implementing it rather than amending `container.ContainerWaitOKBody` with an additional field of type `error`, and then returning it.

Alternatively, leave it as is and implement `container.WaitConditionRemovedOrError` (for the sake of a better name) which will behave as above.

**Comment by kolyshkin** [comment]
OK here's a slightly less naive approach, following the first idea from my previous message. The idea is: if a container removal failed, it doesn't make sense to make the client wait for WaitConditionRemoved forever -- instead we report an error back, so a client can retry removal or at least report an error instead of being stuck.

Here's the patch.

### engine:
```diff
diff --git a/api/server/router/container/container_routes.go b/api/server/router/container/container_routes.go
index 30fd3a1..53a50e5 100644
--- a/api/server/router/container/container_routes.go
+++ b/api/server/router/container/container_routes.go
@@ -317,8 +317,14 @@ func (s *containerRouter) postContainersWait(ctx context.Context, w http.Respons
 	// Block on the result of the wait operation.
 	status := <-waitC
 
+	waitError := ""
+	if status.Err() != nil {
+		waitError = status.Err().Error()
+	}
+
 	return json.NewEncoder(w).Encode(&container.ContainerWaitOKBody{
 		StatusCode: int64(status.ExitCode()),
+		Error:      waitError,
 	})
 }
 
diff --git a/api/swagger.yaml b/api/swagger.yaml
index 75276c1..b1b9125 100644
--- a/api/swagger.yaml
+++ b/api/swagger.yaml
@@ -5701,6 +5701,10 @@ paths:
                 description: "Exit code of the container"
                 type: "integer"
                 x-nullable: false
+              Error:
+                description: "Container removal error, if any"
+                type: "string"
+                x-nullable: false
         404:
           description: "no such container"
           schema:
diff --git a/api/types/container/container_wait.go b/api/types/container/container_wait.go
index 77ecdba..609e1dd 100644
--- a/api/types/container/container_wait.go
+++ b/api/types/container/container_wait.go
@@ -11,6 +11,10 @@ package container
 // swagger:model ContainerWaitOKBody
 type ContainerWaitOKBody struct {
 
+	// Container removal error, if any
+	// Required: true
+	Error string `json:"Error"`
+
 	// Exit code of the container
 	// Required: true
 	StatusCode int64 `json:"StatusCode"`
diff --git a/client/container_wait.go b/client/container_wait.go
index 854c6c0..712c616 100644
--- a/client/container_wait.go
+++ b/client/container_wait.go
@@ -50,6 +50,10 @@ func (cli *Client) ContainerWait(ctx context.Context, containerID string, condit
 			errC <- err
 			return
 		}
+		if res.Error != "" {
+			errC -< error(res.Error)
+			return
+		}
 
 		resultC <- res
 	}()
diff --git a/daemon/delete.go b/daemon/delete.go
index 99b515d..078ad8a 100644
--- a/daemon/delete.go
+++ b/daemon/delete.go
@@ -119,12 +119,18 @@ func (daemon *Daemon) cleanupContainer(container *container.Container, forceRemo
 		metadata, err := daemon.stores[container.Platform].layerStore.ReleaseRWLayer(container.RWLayer)
 		layer.LogReleaseMetadata(metadata)
 		if err != nil && err != layer.ErrMountDoesNotExist && !os.IsNotExist(errors.Cause(err)) {
-			return errors.Wrapf(err, "driver %q failed to remove root filesystem for %s", daemon.GraphDriverName(container.Platform), container.ID)
+			e := errors.Wrapf(err, "driver %q failed to remove root filesystem for %s", daemon.GraphDriverName(container.Platform), container.ID)
+			container.SetError(e)
+			container.SetRemoved()
+			return e
 		}
 	}
 
 	if err := system.EnsureRemoveAll(container.Root); err != nil {
-		return errors.Wrapf(err, "unable to remove filesystem for %s", container.ID)
+		e := errors.Wrapf(err, "unable to remove filesystem for %s", container.ID)
+		container.SetError(e)
+		container.SetRemoved()
+		return e
 	}
 
 	linkNames := daemon.linkIndex.delete(container)
```

### client (i.e. docker-cli, excluding vendoring for new api/types/container):
```diff
diff --git a/cli/command/container/utils.go b/cli/command/container/utils.go
index 024e6ed..1f4fcfc 100644
--- a/cli/command/container/utils.go
+++ b/cli/command/container/utils.go
@@ -37,6 +37,9 @@ func waitExitOrRemoved(ctx context.Context, dockerCli *command.DockerCli, contai
        go func() {
                select {
                case result := <-resultC:
+                       if result.Error != "" {
+                               logrus.Errorf("Error from engine: %v", result.Error)
+                       }
                        statusC <- int(result.StatusCode)
                case err := <-errC:
                        logrus.Errorf("error waiting for container: %v", err)
```

Now, with this change in engine, the old client is unstuck, which actually fixes the bug (but does not report an error), and the new (patched) client reports an error.

This is how it looks with a patched client:
```
[root@kir-ce73-gd 34672]# ./repro21.sh 
...
Iter:   100 Jobs:    99
real	0m3.622s
user	0m0.142s
sys	0m0.200s
Waiting...
ERRO[0009] Error from engine: driver "overlay2" failed to remove root filesystem for 5e0a03161816ed0e1923506981069e1989448a379eadd989298be0e49ea00861: remove /var/lib/docker/overlay2/f512ace537622f7fb1b6c47aae3ce9099aa70d35ea3f63c594726c82ff3eef06/merged: device or resource busy 

real	0m9.525s
user	0m1.826s
sys	0m1.280s
```

I am thinking about amending the client to retry removing container.

@thaJeztah @cpuguy83 @jlhawn what do you think? Would it be easier to review if I create a PR?

**Comment by thaJeztah** [comment]
> I am thinking about amending the client to retry removing container.

Hm, wondering if we should do that, or just fail (just from that description, I should dive into the changes in your patch).

Basically, on older daemons/API versions, the client was responsible for removing the container if;

- the container was running in the foreground (i.e., client kept attached)
- after the container exited, the client would do a `docker rm` for the container

There are various situations where this went wrong, e.g., if the connection dropped or the client was terminated, the container would not be removed. I assume in the old situation (old client, old daemon, with client-side removal), a container that failed to be removed would also produce an error, and no retry was done. Basically, the "client-side auto remove" has always been on a "best effort" base, and it's known to be "hit and miss" for some cases. Given that the daemon-side auto removal replaces that feature, I'm not sure how much improvements should be made to that feature (other than of course making it work the same as it worked with older daemons/clients)

**Comment by kolyshkin** [comment]
@thaJeztah I tend to agree, no need to retry from the client. The only remaining question is whether docker cli should return with 0 or 1 in this very case when `docker run` succeed but `--rm` part of it did not.

Yet another more radical approach would be to not wait for container removal at all in docker-cli as naturally it might never succeed, but do it asynchronously (i.e. ask for removal but don't wait for it).

Anyway, I'll open a PR with the code that I have now, and we'll go from there.

**Comment by thaJeztah** [comment]
@kolyshkin for the `0` or `1`, we should check the behaviour of docker 1.11 (?) or whatever version had the remove client-side.

**Comment by kolyshkin** [comment]
As a separate note, a daemon might retry the container removal if it has failed. An analogy would be `EAGAIN` or `EBUSY` returned from some Linux system calls -- we can return something like this from say a graph driver, in which case the daemon is supposed to retry the removal.
